demos/source/aws_iot_demo_jobs.c

```diff
@@ -0,0 +1,618 @@
+
+        /* AWS IoT recommends the use of the Thing Name as the MQTT client ID. */
+        connectInfo.pClientIdentifier = pIdentifier;
+        connectInfo.clientIdentifierLength = ( uint16_t ) strlen( pIdentifier );
+
+        IotLogInfo( "Thing Name is %.*s (length %hu).",
+                    connectInfo.clientIdentifierLength,
+                    connectInfo.pClientIdentifier,
+                    connectInfo.clientIdentifierLength );
+
+        /* Establish the MQTT connection. */
+        connectStatus = IotMqtt_Connect( &networkInfo, &connectInfo, TIMEOUT_MS, pMqttConnection );
+
+        if( connectStatus != IOT_MQTT_SUCCESS )
+        {
+            IotLogError( "MQTT CONNECT returned error %s.", IotMqtt_strerror( connectStatus ) );
+
+            status = EXIT_FAILURE;
+        }
+    }
+
+    return status;
+}
+
+int RunJobsDemo( bool awsIotMqttMode,
+                 const char * pIdentifier,
+                 void * pNetworkServerInfo,
+                 void * pNetworkCredentialInfo,
+                 const IotNetworkInterface_t * pNetworkInterface )
+{
+    int32_t status = EXIT_SUCCESS;
+    IotMqttConnection_t mqttConnection = IOT_MQTT_CONNECTION_INITIALIZER;
+    AwsIotJobsCallbackInfo_t callbackInfo = AWS_IOT_JOBS_CALLBACK_INFO_INITIALIZER;
+    AwsIotJobsError_t err = AWS_IOT_JOBS_SUCCESS;
+    size_t thingNameLength = 0;
+
+    bool initialized = false;
+    bool connected = false;
+
+    callbackInfo.function = _jobsCallback;
+    callbackInfo.pCallbackContext;
+
+    if( status == EXIT_SUCCESS )
+    {
+        thingNameLength = strlen( pIdentifier );
+
+        if( thingNameLength == 0 )
+        {
+            IotLogError( "The length of the Thing Name (identifier) must be nonzero." );
+            status = EXIT_FAILURE;
+        }
+    }
+
+    if( status == EXIT_SUCCESS )
+    {
+        status = _initDemo();
+        initialized = status == EXIT_SUCCESS;
+    }
+
+    if( status == EXIT_SUCCESS )
+    {
+        status = _establishMqttConnection(
+            pIdentifier, pNetworkServerInfo, pNetworkCredentialInfo, pNetworkInterface, &mqttConnection );
+        connected = status == EXIT_SUCCESS;
+    }
+
+    if( status == EXIT_SUCCESS )
+    {
+        err = AwsIotJobs_SetNotifyNextCallback( mqttConnection, pIdentifier, thingNameLength, 0, &callbackInfo );
+
+        status = err == AWS_IOT_JOBS_SUCCESS ? EXIT_SUCCESS : EXIT_FAILURE;
+    }
+
+    if( status == EXIT_SUCCESS )
+    {
+        IotLogInfo( "--- Add Job using AWS CLI --- \r\n" ); /* Add an extra line for emphasis. */
+
+        for( ; ; )
+        {
+            if( Atomic_CompareAndSwap_u32( &finishFlag, 0, JOBS_DEMO_FINISHED ) == 1 )
+            {
+                break;
+            }
+
+            IotClock_SleepMs( 100 );
+        }
+    }
+
+    IotClock_SleepMs( 1000 );
+
+    if( status == EXIT_SUCCESS )
+    {
+        err = AwsIotJobs_SetNotifyNextCallback( mqttConnection, pIdentifier, thingNameLength, 0, NULL );
```

Comment:
This function has changed.  Please see the latest code (should not pass NULL as the last parameter).